requirements.txt

```diff
@@ -6,3 +6,4 @@ biopandas==0.5.1.dev0
 scipy>=1.8.0
 matplotlib>=3.5.1
 tqdm>=4.57.0
+cmake>=3.22.1
```

Comment:
I think we should add a comment here to note that this dependency is only required for testing, in case it might be an issue for some specific systems of our users

Reply:
Can do.

As a general comment, `requirements.txt` isn't the right tool to document the dependencies of two different audiences, in this case end users vs. developers. I wasn't even sure if we should even mention CMake at all here, since `requirements.txt` is used by PyPI to document user requirements, and they will never have to call CTest or Pylint when running pyMBE simulations. But we still need to document CMake and Pylint version requirements somewhere for our contributors. We could do that in the readme file, but this can quickly become outdated.

This is a common issue for Python package maintainers, and I don't have a satisfying solution at hand. Maintaining two `requirements.txt` files is discouraged due to the cybersecurity threat it poses (someone once registered a faulty Python package [`requirements-dev.txt`](https://pypi.org/project/requirements-dev/) as a joke to raise awareness, although now this package raises a meaningful error message). In the ESPResSo user guide we use the requirement file as a constraint file with syntax `pip3 install -c requirements.txt numpy scipy jupyterlab` instead of `-r`, but then you need to specify the subset of Python packages you want to install.

I also quickly went through the [`poetry`](https://python-poetry.org/) tutorial, since it was recommended to me as a potential solution, and while it solves the issue of sorting packages into hard and soft dependencies using a tree structure with optional leaves, under the hood every change to the text file triggers and installation of the corresponding package in a `virtualenv` environment. That's because poetry is primarily designed for packaging and publishing projects, which is a lot more than what we are asking for, and can be a bit confusing to contributors who are only interested in quickly setting up a `venv` or `conda` environment. I see a lot of value in poetry, however I am not fully convinced this is the right tool for us, at this specific point in time.

Here is a minimal working example I ran in a Docker container; the `@^1.23` and `@~7.4.4` labels generate `>=1.23,<2.0` resp. `>=7.4.4,<7.5` version ranges, but these are processed internally to yield a pip input file with frozen version numbers that satisfy these ranges:

```sh
pip3 install --user poetry poetry-plugin-export
poetry new pyMBE
cd pyMBE
poetry add --group dev 'pylint@3.0.3' 'coverage@~7.4.4' 'cmake@^3.22.1'
poetry add --group docs 'pdoc@^14.3'
poetry add 'numpy@^1.23'
poetry lock
poetry export -f requirements.txt --without-hashes --only dev -o only-dev.txt
pip3 install --user -r only-dev.txt
cat pyproject.toml
cat only-dev.txt
```
```toml
[tool.poetry]
name = "pyMBE"
version = "0.8.0"
description = "Python-based Molecule Builder for ESPResSo"
authors = ["Your Name <you@example.com>"]
readme = "README.md"

[tool.poetry.dependencies]
python = "^3.8"
numpy = "^1.23"

[tool.poetry.group.dev]
optional = true

[tool.poetry.group.dev.dependencies]
pylint = "3.0.3"
coverage = "~7.4.4"
cmake = "^3.22.1"

[tool.poetry.group.docs]
optional = true

[tool.poetry.group.docs.dependencies]
pdoc = "^14.3"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
```
```
astroid==3.0.3 ; python_version >= "3.8" and python_version < "4.0"
cmake==3.30.2 ; python_version >= "3.8" and python_version < "4.0"
colorama==0.4.6 ; python_version >= "3.8" and python_version < "4.0" and sys_platform == "win32"
coverage==7.4.4 ; python_version >= "3.8" and python_version < "4.0"
dill==0.3.8 ; python_version >= "3.8" and python_version < "4.0"
isort==5.13.2 ; python_version >= "3.8" and python_version < "4.0"
mccabe==0.7.0 ; python_version >= "3.8" and python_version < "4.0"
platformdirs==4.2.2 ; python_version >= "3.8" and python_version < "4.0"
pylint==3.0.3 ; python_version >= "3.8" and python_version < "4.0"
tomli==2.0.1 ; python_version >= "3.8" and python_version < "3.11"
tomlkit==0.13.2 ; python_version >= "3.8" and python_version < "4.0"
typing-extensions==4.12.2 ; python_version >= "3.8" and python_version < "3.11"
```